src/Workspaces/Core/Portable/Storage/AbstractPersistentStorageService.cs

```diff
@@ -41,57 +37,58 @@ protected AbstractPersistentStorageService(IPersistentStorageConfiguration confi
 
-    public ValueTask<IChecksummedPersistentStorage> GetStorageAsync(SolutionKey solutionKey, CancellationToken cancellationToken)
+    public async ValueTask<IChecksummedPersistentStorage> GetStorageAsync(SolutionKey solutionKey, CancellationToken cancellationToken)
     {
-        return solutionKey.FilePath == null
-            ? new(NoOpPersistentStorage.GetOrThrow(Configuration.ThrowOnFailure))
-            : GetStorageWorkerAsync(solutionKey, cancellationToken);
-    }
+        if (solutionKey.FilePath == null)
+            return NoOpPersistentStorage.GetOrThrow(solutionKey, Configuration.ThrowOnFailure);
+
+        // Without taking the lock, see if we can use the last storage system we were asked to create.  Ensure we use a
+        // using so that if we don't take it we still release this reference count.
+        using var existing = _currentPersistentStorage?.TryAddReference();
+        if (solutionKey == existing?.Target.SolutionKey)
+        {
+            // Success, we can use the current storage system.  Ensure we increment the reference count again, so that
+            // this stays alive for the caller when the above reference count drops.
+            return PersistentStorageReferenceCountedDisposableWrapper.AddReferenceCountToAndCreateWrapper(existing);
+        }
+
+        var workingFolder = Configuration.TryGetStorageLocation(solutionKey);
+        if (workingFolder == null)
+            return NoOpPersistentStorage.GetOrThrow(solutionKey, Configuration.ThrowOnFailure);
 
-    internal async ValueTask<IChecksummedPersistentStorage> GetStorageWorkerAsync(SolutionKey solutionKey, CancellationToken cancellationToken)
-    {
         using (await _lock.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
         {
-            // Do we already have storage for this?
-            if (solutionKey.Id == _currentPersistentStorageSolutionId)
+            // See if another thread set to the solution we care about while we were waiting on the lock.
+            using var current = _currentPersistentStorage?.TryAddReference();
+            if (solutionKey != current?.Target.SolutionKey)
             {
-                // We do, great. Increment our ref count for our caller.  They'll decrement it
-                // when done with it.
-                return PersistentStorageReferenceCountedDisposableWrapper.AddReferenceCountToAndCreateWrapper(_currentPersistentStorage!);
+                // We either don't have a current storage, or the current storage was pointing at some other solution. In
+                // either case, we want to create a new storage instance and point at that.  If the current storage was
+                // point at some other solution, then lower its ref count (by calling DisposeAsync, outside of the lock) to
+                // indicate that *we* are letting go of it.
+                _ = DisposeStorageAsync(_currentPersistentStorage);
```

Comment:
>_ = DisposeStorageAsync(_currentPersistentStorage); [](http://example.com/codeflow?start=16&length=51)

This is tricky. Would it be a bit less tricky if we just did the DisposeAsync call explicitly outside the using if needed? Something like:

ReferenceCountedDisposable<IChecksummedPersistentStorage>? storageRequiringDisposal = null;
IChecksummedPersistentStorage wrapper;

using (await ...)
{
  ...
  if (solutionKey != current?.Target.SolutionKey)
  {
    storageRequiringDisposal  = _currentPersistantStorage;
    _currentPersistentStorage = ...
  }

  Contract.ThrowIfNull(_currentPersistentStorage);
  wrapper = PersistentStorageReferenceCountedDisposableWrapper.AddReferenceCountToAndCreateWrapper(_currentPersistentStorage);
}

if (storageRequiringDisposal != null)
  await storageRequiringDisposal.DisposeAsync().ConfigureAwait(false);

return wrapper;

Reply:
takea  look :)

Reply:
i started there, but it was more complex (especially ensuring that we  returned the wrapper, even in hte presence of a throw).

Reply:
talked offline on this.